### PR TITLE
Add WeakKeyIdDict

### PR DESCRIPTION
**Pull request by fingolfin**
- Added data-structure WeakKeyIdDict
- Sync WeakKeyIdDict with latest Julia WeakKeyDict implementation
- Fix some import issues in WeakValueDict

See https://github.com/oscar-system/Oscar.jl/pull/1872

CC @mauro3

### CODE REVIEW

**Comment by fingolfin** [comment]
Good point -- I only copied the existing tests, but of course syncing with master.

So I've added some tests, and of course this uncovered various bugs 😂 . I've also tried to add a test for the GC collecting the weak refs, but I can't get it to reliably work, so it is for now disabled.

I'll wait for the updated coverage report to see where else I can add some more.

File: test/WeakKeyIdDict-test.jl [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  wkd[tmp] = 2
+  @test length(wkd) == 2
+end
+# at this point there is no strong reference left to the vector [42]
+# previously reachable via tmp
+GC.gc(true)
+
+@test length(wkd) == 1
+@test length(keys(wkd)) == 1
+@test WeakKeyIdDict(A => 1) == wkd
```
Comment (fingolfin) [inline test/WeakKeyIdDict-test.jl:157]: Assuming the CI tests pass now, I am fine with how high the coverage now is (> 90%). Note this ugliness: I could not find a way to get those tests which interact with the GC to work "right" while inside a `@testset`. But I am not aware of a serious drawback of the above, other than it not being reported in the output of `Pkg.test("AbstractAlgbera")` ... ?

Comment (lgoettgens) [reply test/WeakKeyIdDict-test.jl:157]: Does it work if you put it into a `let` scope? Currently, CI fails as there are two global objects both named `A` in the tests.

**Comment by ericphanson** [comment]
I wonder if WeakKeyIdDict could do in a separate WeakKeyIdDicts.jl package, so it could be reused elsewhere easier?

**Comment by ericphanson** [comment]
BTW @haberdashPI has created this package here: https://github.com/beacon-biosignals/WeakKeyIdDicts.jl

**Pull request merged**